### ISSUE TITLE
Need insight into (pathological?) colorization behavior on Azure DevOps

### ISSUE THREAD

**Issue by bskinn** [fish-shell/fish-shell#5736]
I'm working on pypa/virtualenv#1330, which is trying to put under test the prompt modification behavior for the various shells supported by virtualenv, including fish. I'm seeing a curious platform-dependent problem with some fish code that looks to be a quirk of the way Azure DevOps sets up their bash terminals.

The specific fish code that's the origin of my problem is from [here](https://github.com/pypa/virtualenv/blob/94c6bcb4a3f8d7ff63e2215b18d76514d147a762/virtualenv_embedded/activate.fish#L92):

```
function fish_prompt
    ...
    printf '%s(%s) ' (set_color normal) (basename "$VIRTUAL_ENV")
    ...
end
```

-----

**[GOOD BEHAVIOR]** When I run the test covering this piece of code from (a putty session into) my local dev box (Debian stretch, fish v2.4.0, `echo $TERM` yields `linux`, `tput colors` yields `8`), my test finds "`(env) `" as the prefix to the prompt, as I would expect.  Also:

```
> set_color normal | od -t x1z
0000000 1b 5b 33 30 6d 1b 5b 6d 0f                       >.[30m.[m.<
0000011
```

Seems fine to me.

-----

**[BAD BEHAVIOR]** However, when the test suite runs in Azure DevOps (Ubunto Xenial, fish 2.2.0-3), it appears that the terminal type is such that `set_color normal` returns *absolutely nothing at all*, null content. If I inject `set_color normal | od -t x1z` into my test script the output is just:

```
0000000
```
Probably important: I'm seeing multiple `set_color: Could not set up terminal` messages being captured at `stderr`. So... I'm guessing that Azure's terminal setup is pathological in some way, in that `set_color` is throwing an error when called, rather than returning a sequence of ANSI codes?

In any event, this is giving me a spurious test failure on Azure, because the null return of `(set_color normal)` is munging the `printf` in the above code, leading to a prompt prefix of "`env() `" instead of the correct "`(env) `" as on my Debian box. 

Weirdly, if I'm reading the outputs correctly, both `tput colors` and `echo $TERM` print ***empty strings*** to stdout when injected into my fish test scripts. This implies to me that Azure simply *hasn't defined* their terminal type? I don't see how this is possible, though, as if I try to `env TERM= fish` or `env TERM="" fish` or `env TERM=" " fish` on my Debian box, fish complains and falls back to `ansi`.

Any advice on how to demunge this Azure terminal setup?  Neither a `set TERM ansi-mono` nor a `set TERM xterm-mono` fixed the problem when added to the preamble of the fish script executed by the test function.

**Comment by faho** [comment]
The problem here is that `set_color normal` in Azure outputs absolutely _nothing_, because it doesn't have any color.

Empty expansions cancel out adjacent text - like e.g. brace expansions "multiply" it.

The solution would be for us to print at least a newline - #5443.

The workaround for you is to save the `set_color` sequence in a variable and quote that:

```fish
set -l normal (set_color normal)
printf '%s(%s) ' "$normal" (basename "$VIRTUAL_ENV")
```

(even with empty output, `"$normal"` will expand to one empty argument, so the printf then uses it for the first `%s`)

or set $TERM to an entry that exists and has at least 1 color.

I'm closing this as a duplicate of #5443.

**Comment by bskinn** [comment]
Thanks very much for the info! Sorry for the duplicate. 😞

**Comment by faho** [comment]
>To highlight, though: I'd tried, e.g., set TERM ansi-mono in the preamble of the fish script, but that didn't seem to work.

Try `set -gx TERM ansi-mono`.

**Comment by bskinn** [comment]
Hmm, nope, still getting `set_color: Could not set up terminal`. :-(

Neither `set -gx TERM ansi-mono` as the first command of the fish script, nor executing the script with `TERM=ansi-mono` preloaded in the environment fixes it.

I have a tweet out to ADO about terminal configuration -- maybe they'll have some suggestions....

**Comment by faho** [comment]
>Neither set -gx TERM ansi-mono as the first command of the fish script, nor executing the script with TERM=ansi-mono preloaded in the environment fixes it.

Then `ansi-mono` most likely does not exist as a terminfo entry on that system. It's in the [ncurses-term](https://packages.debian.org/stretch/all/ncurses-term/filelist) package, so either install that or pick one of the definitions in [ncurses-base](https://packages.debian.org/stretch/all/ncurses-base/filelist).

`dumb` should work (and has no color), though you probably just want `xterm-256color` to actually test the color values?

**Comment by bskinn** [comment]
No, I don't care about the color values -- I just need set_color to return *some*thing non-NULL on ADO so that the `printf` sees the `set_color` result as one of its arguments, and thus puts the env name inside, rather than before, the parentheses.

I'd like to avoid having to rewrite the fish script with the `set -l normal` etc. as you indicated above, just to cope with an idiosyncracy of the CI... but I think that should work fine if it comes down to it.

Thanks again for the help -- I had no idea where to look for a list of terminals, those links are a huge help!

**Comment by faho** [comment]
>No, I don't care about the color values -- I just need set_color to return something non-NULL on ADO so that the printf sees the set_color result as one of its arguments

That means you need it to have a $TERM with color. Which means `xterm` or related is as good as any, and `dumb` won't work (just checked again, it triggers the issue, even if it won't print an error).

For it to work without color, you'll have to wait for a patch to fish (and it won't be to 2.2.0 or 2.4.0, those versions are quite old).

**Comment by bskinn** [comment]
<nod>, I checked `dumb` myself manually on my Debian box and confirmed that `set_color` puts out a null output in that case. Definitely not a good option.

It works now: Injecting `TERM=linux` into the surrounding environment used to execute the fish script (via Python's `subprocess.call(..., env=ENVIRON)` machinery) worked like a charm. 👍  It also probably helped that I actually injected it this time within the correct test function. 🤦‍♂️  So, `TERM=ansi-mono` might well have worked, if I hadn't put it in the wrong place the first time.

Owe you a drink of whatever your preference, @faho, should we ever meet in person!
